### DISCUSSION TITLE
PoC with OpenAirInterface (OAI) SPGWu

### DISCUSSION THREAD

**Discussion by navarrothiago** [navarrothiago/upf-bpf#52] [Ideas]
Because the upf-bpf is a C++ library, a good candidate to do this integration is the OpenAirInterface's SPGWu, which is also implemented using C++. Here is some suggestion of task for this PoC:

- [ ]  Deploy the [OAI spgwu](https://github.com/OPENAIRINTERFACE/openair-spgwu-tiny)
- [ ]  Create/Find a component to send the PFCP messages to spgwu [1][2]
- [ ]  Test OAI with the setup below
- [ ]  Compile the OAI with upf-bpf library
- [ ]  Test OAI + upf-bpf with the same setup

[1] -  I didnt find any open source tool that could be a PFPC client (e.g. [here](https://www.synopsys.com/software-integrity/security-testing/fuzz-testing/defensics/protocols/pfcp-client.html)).
[2] - I have some hardcoded implementation. Check [here](https://github.com/navarrothiago/openair-cn-cups/blob/develop-spgwu/test/integration/spgwu_tests.cpp#L140-L210). The idea is to inject the ITTI messages in SPGWu. The ITTI is a m

The test case description is defined [here](https://github.com/navarrothiago/openair-cn-cups/issues/1).

![image](https://user-images.githubusercontent.com/42647168/129460397-391b4918-9816-46f2-b85b-6081979e8864.png)

**Comment by infinitydon** [comment]
What are the steps to  **_Compile the OAI with upf-bpf library_**? As OAI code-base is constantly in development, is there any particular release/commit/branch to use for the SPGWu?

**Reply by navarrothiago** [reply]
We could target one version. I am opened for suggestion. Today, I dont have the steps to compile with OAI. As I mentioned, I have already done a simple integration, but it is a little messy to be honest. One starting pointing is to check the `develop-spgwu` branch in this repo - https://github.com/navarrothiago/openair-cn-cups. WDYT? Basically, the idea is to modify the CMakeLists.txt to include this library. I remember that I faced a lot of issues an took some time to solved it.

**Reply by infinitydon** [reply]
I have not been able to find time to try this out but as regards sending PFCP association request the following opensource might be helpful:

https://github.com/TheWayYouMakeMeFeel/pfcp-kitchen-sink

**Reply by navarrothiago** [reply]
Ohw!  This is nice! Thanks for sharing. I was looking for this sometime ago. With this tool, we'd replace the `SessionControlTester` to the `pfcp-kichen-sink`. It would be easier to test with the OAI SPGWu, we would not know the internals. We should concern about the compatibility between the PFCP OAI implementation and the PFCP of this repo. It should be compatible... but, you know :smile:. One task that I have in mind is to create the PFCP client using https://github.com/wmnsk/go-pfcp. This library seems to be more mature.
